rinja_derive/src/generator.rs

```diff
@@ -1056,128 +1053,89 @@ impl<'a> Generator<'a> {
                     size_hint += s.len();
+                    let mut items = vec![s];
+                    while let Some((_, Writable::Lit(s))) = it.peek() {
+                        size_hint += s.len();
+                        items.push(s);
+                        it.next();
+                    }
+                    lines.writeln(format_args!(
+                        "writer.write_str({:#?})?;",
+                        items.as_slice().join(""),
+                    ));
                 }
                 Writable::Expr(s) => {
+                    size_hint += 3;
+
                     let mut expr_buf = Buffer::new();
-                    let wrapped = self.visit_expr(ctx, &mut expr_buf, s)?;
-                    let cacheable = is_cacheable(s);
-                    size_hint += self.named_expression(
-                        &mut buf_expr,
-                        &mut buf_format,
-                        expr_buf.buf,
-                        wrapped,
-                        cacheable,
-                        &mut expr_cache,
-                    );
+                    let expr = match self.visit_expr(ctx, &mut expr_buf, s)? {
+                        DisplayWrap::Wrapped => expr_buf.buf,
+                        DisplayWrap::Unwrapped => format!(
+                            "(&&{CRATE}::filters::AutoEscaper::new(&({}), {})).\
+                                rinja_auto_escape()?",
+                            expr_buf.buf, self.input.escaper,
+                        ),
+                    };
+                    let idx = if is_cacheable(s) {
+                        match expr_cache.entry(expr) {
+                            Entry::Occupied(e) => *e.get(),
+                            Entry::Vacant(e) => {
+                                buf.writeln(format_args!("&({}),", e.key()));
+                                targets.writeln(format_args!("expr{idx},"));
+                                e.insert(idx);
+                                idx
+                            }
+                        }
+                    } else {
+                        buf.writeln(format_args!("&({expr}),"));
+                        targets.write(format_args!("expr{idx}, "));
+                        idx
+                    };
+                    lines.writeln(format_args!(
+                        "(&&{CRATE}::filters::Writable(expr{idx})).rinja_write(writer)?;",
+                    ));
                 }
             }
         }
-        Ok(WriteParts {
-            size_hint,
-            buffers: Some(WritePartsBuffers {
-                format: buf_format,
-                expr: Some(buf_expr),
-            }),
-        })
-    }
-
-    fn named_expression(
-        &mut self,
-        buf_expr: &mut Buffer,
-        buf_format: &mut Buffer,
-        expr: String,
-        wrapped: DisplayWrap,
-        cacheable: bool,
-        expr_cache: &mut HashMap<String, usize>,
-    ) -> usize {
```

Comment:
I inlined this function. In my opinion the new functions is small enough by itself.